mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+            params, hidden_state, (obs, term_or_trunc), eps
+        )
+
+        # Random key splitting
+        new_key, explore_key = jax.random.split(key, 2)
+
+        action = eps_greedy_dist.sample(seed=explore_key)
+        action = action[0, ...]  # (1, B, A) -> (B, A)
+
+        # repack new selection params
+        next_action_selection_state = ActionSelectionState(
+            params, next_hidden_state, t + cfg.arch.num_envs, new_key
+        )
+        return next_action_selection_state, action
+
+    def action_step(action_state: ActionState, _: Any) -> Tuple[ActionState, Dict]:
+        """Selects an action, steps global env, stores timesteps in global rb and repacks the
+        parameters for the next step.
+        """
+
+        action_selection_state, env_state, buffer_state, obs, terminal, term_or_trunc = action_state
+
+        next_action_selection_state, action = select_eps_greedy_action(
+            action_selection_state, obs, term_or_trunc
+        )
+
+        next_env_state, next_timestep = jax.vmap(env.step)(env_state, action)
+
+        # Get reward
+        reward = jnp.mean(
+            next_timestep.reward, axis=-1, keepdims=True
+        )  # NOTE: Combine agent rewards, since QMIX is cooperative.
+
+        transition = Transition(
+            obs, action, reward, terminal, term_or_trunc, next_timestep.extras["real_next_obs"]
+        )
+        # Add dummy time dim
+        transition = tree.map(lambda x: x[:, jnp.newaxis, ...], transition)
+        next_buffer_state = rb.add(buffer_state, transition)
+
+        next_obs = next_timestep.observation
+        # Make compatible with network input and transition storage in next step
+        next_terminal = (1 - next_timestep.discount[..., 0, jnp.newaxis]).astype(bool)
+        next_term_or_trunc = next_timestep.last()[..., jnp.newaxis]
+
+        new_act_state = ActionState(
+            next_action_selection_state,
+            next_env_state,
+            next_buffer_state,
+            next_obs,
+            next_terminal,
+            next_term_or_trunc,
+        )
+
+        return new_act_state, next_timestep.extras["episode_metrics"]
+
+    def prep_inputs_to_scannedrnn(obs: Observation, term_or_trunc: chex.Array) -> chex.Array:
+        """Prepares the inputs to the RNN network for either getting q values or the
+        eps-greedy distribution.
+
+        Mostly swaps leading axes because the replay buffer outputs (B, T, ... )
+        and the RNN takes in (T, B, ...).
+        """
+        hidden_state = ScannedRNN.initialize_carry(
+            (cfg.system.sample_batch_size, obs.agents_view.shape[2]), cfg.network.hidden_state_dim
+        )
+        # the rb outputs (B, T, ... ) the RNN takes in (T, B, ...)
+        obs = switch_leading_axes(obs)  # (B, T) -> (T, B)
+        term_or_trunc = switch_leading_axes(term_or_trunc)  # (B, T) -> (T, B)
+        obs_term_or_trunc = (obs, term_or_trunc)
+
+        return hidden_state, obs_term_or_trunc
```

Comment:
Could move this to `utils/rnn` or `utils/qlearning` because it's used in both Q learning systems

Reply:
I know we reuse the code a bit between the two. But since this accessed the config without having it passed in, it takes a bit of work to move it to util file. If you feel strongly, I am happy to do so?

Reply:
A way around this is to pass in the config every time we call this, but it feels a bit extra to me? Or we need a function that inits this which also feels a bit extra to me.

Reply:
I see, ye passing the config in would be better, but I think it's fine :smile: